### ISSUE TITLE
Inclusion of Google Fonts by default might create a legal hazard due to GDPR

### ISSUE THREAD

**Issue by gadmm** [ocaml/odoc#694]
The default `odoc.css` links to https://fonts.googleapis.com. At https://complianz.io/google-fonts-and-gdpr-does-it-work/ and elsewhere, there are arguments that this is against EU's GDPR. Google Fonts should either be blocked by the webpage until consent is given by the user, or (simpler and more realistic) the fonts should be self-hosted, they say. (Whether this affects only European users of odoc, or all users of odoc serving to clients in the EU, frankly flies over my head.)

Users can already change `odoc.css` to point to self-hosted fonts, but they are probably unaware of the legal hazards of the default settings. In order to err safely on the side of “legal by default”,  it would be nice if:
- the inclusion of remote google fonts was opt-in,
- `odoc` assisted in self-hosting the fonts.

**Comment by gadmm** [comment]
Here is an example where the remote css is replaced with self-hosted fonts, this is fairly simple: https://guillaume.munch.name/software/ocaml/odoc.css. Tools like https://google-webfonts-helper.herokuapp.com/fonts have been helpful to generate the css and to download the font files.

**Comment by Julow** [comment]
I completely agree !

There's a few more reasons to avoid this:
- Google might be using requests made to the fonts to invade privacy (I mean, they certainly do but are only getting a small amount of data).
- The generated documentation is not 100% self contained because it changes its appearance whether an internet connection is available or not.
- Some people don't like when webpages that could be simple are not (contain fonts or CSS that can't be removed without breaking the page)

My opinion is that we should not load any font and use basic fonts (eg. no url in the CSS file and `font-family: Arial, sans-serif;`).
Modifying the CSS file is already encouraged by recent changes (not yet documented though) so having a simpler graphic design by default seems better than the opposite.

**Comment by gadmm** [comment]
You might be interested in the experience of using system fonts from the following website: https://meta.stackexchange.com/questions/364048/we-are-switching-to-system-fonts-on-may-10-2021 (it seems to bring unexpected complications).

As a user I would prefer something that just works and is nice out of the box and does not require me looking into customizing CSS.

**Comment by Julow** [comment]
My comment above is a bit out of scope. Odoc comes with a graphic design by default, so using specific fonts is reasonable.

Perhaps the fonts should be distributed with Odoc and with the generated HTML (like the CSS file) ? The 3 fonts are libre (Noticia Text, Fira Mono and Fira Sans).

Unrelated again but perhaps we can reduce the number of different fonts and weights ? There's one for navigation and headings and a different one for the text, there's also 4 different `font-weight` values. That's reasonable for an entire website but I think simpler is better in this case because the design is intended to be personalized.

**Comment by gadmm** [comment]
This issue is critical, it creates a legal hazard for odoc users.

https://www.theregister.com/2022/01/31/website_fine_google_fonts_gdpr/

> a German court fined an unidentified website €100 ($110, £84) for violating EU privacy law by importing a Google-hosted web font.
> [...]
> The ruling directs the website to stop providing IP addresses to Google and threatens the site operator with a fine of €250,000 for each violation, or up to six months in prison, for continued improper use of Google Fonts.

https://rewis.io/urteile/urteil/lhm-20-01-2022-3-o-1749320/

**Comment by dbuenzli** [comment]
> Perhaps the fonts should be distributed with Odoc and with the generated HTML (like the CSS file) ?

FWIW that's what odig does (because it's designed to work offline). Just make sure you add all the font licenses to the repo and mention them in the `opam` file.

See for example

* https://github.com/b0-system/odig/blob/c4727e47700f95fbae1c847017360b0afd6cd0f9/LICENSE.md
* https://github.com/b0-system/odig/blob/c4727e47700f95fbae1c847017360b0afd6cd0f9/opam#L16

**Comment by toastal** [comment]
> My opinion is that we should not load any font and use basic fonts (eg. no url in the CSS file and `font-family: Arial, sans-serif;`).

Arial looks really bad on my machine. I have my chosen font already selected by my user agent as well. I think `font-family: sans-serif`, is better than the proposed one. But that could maybe be improved as `font-family: ui-sans-serif` or `font-family: system-ui` depending on goals (though it appears they want the author serif). I don't think the right thing to do includes adding Microsoft's old fonts in the stack and not prioritizing respecting the user agent and/or the OS.

---

Personally, I use a uBlock Origin rule to block _all_ third-party fonts because 95% of the time, they are coming from a source I don't trust with my data like Google, Adobe, etc. As such, I don't see the fonts, but this mostly causes an issues in the code blocks where `courier` (also renders horribly) is included for some reason when, again, I already have my user agent set with my preferred monospace font already with `font-family: monospace`.
